### ISSUE TITLE
getBalance incorrect output for dead contract.

### ISSUE THREAD

**Issue by BenjaminSchaaf** [paritytech/parity#8563]
I'm running:

- **Version**: Parity/v1.9.7-unstable-7919629-20180423/x86_64-linux-gnu/rustc1.25.0
- **How installed?**: snap
- **Are you fully synchronized?**: no
- **Which network are you connected to?**: ethereum
- **Did you try to restart the node?**: no
- **Which API are you using**: RPC using Web3js

In block `1605122`, the contract `0x1f78775c8260df084f9a0e5fbdf06487b875ac4d` suicides with the refund address being itself, ie. `0x1f78775c8260df084f9a0e5fbdf06487b875ac4d`. etherscan reports the balance as being `0.0000001 Ether`, but:

```javascript
web3.eth.getBalance('0x1f78775c8260df084f9a0e5fbdf06487b875ac4d', 1605122) => 0
```

The replay and state diff look correct:

```json
    {
      "action": {
        "address": "0x1f78775c8260df084f9a0e5fbdf06487b875ac4d",
        "balance": "0x71afd498d0000",
        "refundAddress": "0x1f78775c8260df084f9a0e5fbdf06487b875ac4d"
      },
      "result": null,
      "subtraces": 0,
      "traceAddress": [
        0,
        0
      ],
      "type": "suicide"
    }
```

```json
    "0x1f78775c8260df084f9a0e5fbdf06487b875ac4d": {
      "balance": {
        "-": "0x38d7ea4c68000"
      },
      "code": {
        "-": "0x3660008037602060003660003473273930d21e01ee25e4c219b63259d214872220a261235a5a03f21560015760206000f3"
      },
      "nonce": {
        "-": "0x0"
      },
      "storage": {}
    },
```

**Comment by Tbaut** [comment]
Can you confirm that you a running a node with pruning archive?
Could you try calling this RPC directly?
`curl --data '{"method":"eth_getBalance","params":["0x1f78775c8260df084f9a0e5fbdf06487b875ac4d","0x187E02"],"id":1,"jsonrpc":"2.0"}' -H "Content-Type: application/json" -X POST localhost:8545`

I do not have any archive node handy but doing a getBalance with the latest block seems to have the same value as what [Etherscan](https://etherscan.io/address/0x1f78775c8260df084f9a0e5fbdf06487b875ac4d) gives:
`curl --data '{"method":"eth_getBalance","params":["0x1f78775c8260df084f9a0e5fbdf06487b875ac4d"],"id":1,"jsonrpc":"2.0"}' -H "Content-Type: application/json" -X POST localhost:8545
`
Result:
`{"jsonrpc":"2.0","result":"0x174876e800","id":1} //100000000000 Wei`

**Comment by BenjaminSchaaf** [comment]
I'm running the node with everything enabled, ie: `parity --tracing on --fat-db on --pruning archive`.

```
$ curl --data '{"method":"eth_getBalance","params":["0x1f78775c8260df084f9a0e5fbdf06487b875ac4d","0x187E02"],"id":1,"jsonrpc":"2.0"}' -H "Content-Type: application/json" -X POST localhost:8545
{"jsonrpc":"2.0","result":"0x0","id":1}
```

**Comment by BenjaminSchaaf** [comment]
I'm also getting the same result with `latest`:
```
$ curl --data '{"method":"eth_getBalance","params":["0x1f78775c8260df084f9a0e5fbdf06487b875ac4d","latest"],"id":1,"jsonrpc":"2.0"}' -H "Content-Type: application/json" -X POST localhost:8545
{"jsonrpc":"2.0","result":"0x174876e800","id":1}
```

**Comment by Tbaut** [comment]
@BenjaminSchaaf the kill function got called at and some more 0.0000001 Ether have been sent at block 3994185. This is what Etherscan is showing as well. I think the question is more: why is the balance 0 if the kill function was called with the same address and should have the reward fund as balance. Am I right?

**Comment by BenjaminSchaaf** [comment]
Yep, that's what I'm trying to ask. Sorry if the reference to etherscan was confusing. The `stateDiff` clearly shows the balance hasn't changed, but `getBalance` returns `0`.

**Comment by sorpaas** [comment]
@BenjaminSchaaf @Tbaut This is the expected behavior for EVM. EVM kills accounts during transaction finalization. So if an account SUICIDE to itself, the fund will be transferred to it, and then killed, effectively reducing the total supply a little bit, but considerably harmless. The additional 0.00..01 Ether is unrelated and is injected in a future block.

However, the tracing log part indeed looks like a bug, I'll take a look on that.

**Comment by sorpaas** [comment]
It looks like our tracer just ignore state changes during transaction finalization. Guess that's the expected behavior right now, but we may want to consider to change it in the future.

**Comment by BenjaminSchaaf** [comment]
That's interesting: I was under the impression the supply of ether couldn't ever go down. Oh well, thanks for the help.
